Add receiver for sapm protocol

codesmith14 opened the pull request.

charless-splunk committed:
Add receiver for sapm protocol

codesmith14 commented:
I think the ci is failing because `sapm-proto` is still private.  @tigrannajaryan could you please make `sapm-proto` public?

owais commented on receiver/sapmreceiver/factory.go:
```diff
@@ -0,0 +1,120 @@
+	// assert config is SAPM config
+	rCfg := cfg.(*Config)
+
+	if rCfg.IsEnabled() {
```
I've never seen this before. _I think_ this is supposed to be used internally by the pipeline and not the factory. If a component is disabled, pipeline should never ask it's factory to create a receiver.

tigrannajaryan replied:
Yes, this should be handled by https://github.com/open-telemetry/opentelemetry-collector/blob/d32cd2a100a49d2dcd34ba4401d50b4ee15adc37/config/config.go#L711
If it is not then we have a bug.

codesmith14 replied:
removed

owais commented on receiver/sapmreceiver/trace_receiver.go:
```diff
@@ -0,0 +1,173 @@
+		for _, err := range errs {
+			fmt.Fprintf(buf, "%s\n", err.Error())
+		}
+		err = errors.New(buf.String())
```
Otel has a package to combine multiple errors: https://github.com/open-telemetry/opentelemetry-collector/blob/master/oterr/errors.go

codesmith14 replied:
cool, I copied that from the jaeger receiver.  I'll update to use that.

codesmith14 replied:
didn't actually have multiple errors any way.  It was a bad copy/paste.  I've removed it entirely.

tigrannajaryan commented on go.sum:
```diff
@@ -191,6 +191,8 @@ github.com/gogo/protobuf v1.2.2-0.20190723190241-65acae22fc9d/go.mod h1:SlYgWuQ5
 github.com/gogo/protobuf v1.2.2-0.20190730201129-28a6bbf47e48/go.mod h1:SlYgWuQ5SjCEi6WLHjHCa1yvBfUnHcTbrrZtXPKa29o=
 github.com/gogo/protobuf v1.3.0 h1:G8O7TerXerS4F6sx9OV7/nRfJdnXgHZu/S/7F2SN+UE=
 github.com/gogo/protobuf v1.3.0/go.mod h1:SlYgWuQ5SjCEi6WLHjHCa1yvBfUnHcTbrrZtXPKa29o=
+github.com/gogo/protobuf v1.3.1 h1:DqDEcV5aeaTmdFBePNpYsp3FlcVH/2ISVVM9Qf8PSls=
```
Is it valid to consume 2 different versions of the same lib? Which one is actually used by "import" statements?

owais replied:
This is not really consuming. It's just recording the checksum for the version so it it sees the version again, it'll reject it if the checksum does not match. Probably some dependency is using 1.3.1 and some other 1.3.0 ?

tigrannajaryan replied:
OK, so this means that according to "Minimal Version Selection" rule 1.3.1 will win.

owais replied:
Actually 1.3.2 wins. Which exact version of a dependency is used is only dictated by go.mod. go.sum only contains checksums. Relevant line: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/48/files#diff-c391b461a7bdb4bc745d488660866ab0R7

tigrannajaryan commented on go.sum:
```diff
@@ -312,6 +314,8 @@ github.com/inconshreveable/mousetrap v1.0.0/go.mod h1:PxqpIevigyE2G7u3NXJIT2ANyt
 github.com/influxdata/influxdb v1.7.7/go.mod h1:qZna6X/4elxqT3yI9iZYdZrWWdeFOOprn86kgg4+IzY=
 github.com/jaegertracing/jaeger v1.14.0 h1:C0En+gfcxf3NsAriMAvQ6LcSFrQ5VQGXddqfty1EpTI=
 github.com/jaegertracing/jaeger v1.14.0/go.mod h1:LUWPSnzNPGRubM8pk0inANGitpiMOOxihXx0+53llXI=
+github.com/jaegertracing/jaeger v1.15.1 h1:7QzNAXq+4ko9GtCjozDNAp2uonoABu+B2Rk94hjQcp4=
```
Same here.

tigrannajaryan commented on go.sum:
```diff
@@ -491,6 +495,8 @@ github.com/shurcooL/vfsgen v0.0.0-20180825020608-02ddb050ef6b/go.mod h1:TrYk7fJV
 github.com/shurcooL/vfsgen v0.0.0-20181202132449-6a9ea43bcacd/go.mod h1:TrYk7fJVaAttu97ZZKrO9UbRa8izdowaMIZcxYMbVaw=
 github.com/signalfx/com_signalfx_metrics_protobuf v0.0.0-20190530013331-054be550cb49 h1:a6us2VYa4abLd4FG6F3BaGEzuq6WIvloIx3M40ePaJ0=
 github.com/signalfx/com_signalfx_metrics_protobuf v0.0.0-20190530013331-054be550cb49/go.mod h1:muYA2clvwCdj7nzAJ5vJIXYpJsUumhAl4Uu1wUNpWzA=
+github.com/signalfx/sapm-proto v0.0.1 h1:At9k1ynR74nmW8VMdfcqCFV9MWq1BdUpvYtyxHB3jeA=
```
I think we need to make this repo public, otherwise this won't build.

tigrannajaryan commented on receiver/sapmreceiver/config.go:
```diff
@@ -0,0 +1,49 @@
+	"github.com/open-telemetry/opentelemetry-collector/config/configmodels"
+)
+
+// Config defines configuration for Jaeger receiver.
```
```suggestion
// Config defines configuration for SAPM receiver.
```

tigrannajaryan commented on receiver/sapmreceiver/config.go:
```diff
@@ -0,0 +1,49 @@
+// IsEnabled returns true if the entity is enabled.
+func (rs *Config) IsEnabled() bool {
+	return !rs.Disabled
+}
```
I think this is not necessary, these all are already implemented in `configmodels.ReceiverSettings`.

codesmith14 replied:
removed

owais commented on receiver/sapmreceiver/trace_receiver.go:
```diff
@@ -0,0 +1,173 @@
+	return func(rw http.ResponseWriter, req *http.Request) {
+
+		// trace this request
+		_, span := trace.StartSpan(sr.defaultCtx, traceSource)
```
Shouldn't this be using the per request context? In fact, I think it should use `req.Context()`.

codesmith14 replied:
This is the main thing I wanted to talk about

tigrannajaryan commented on receiver/sapmreceiver/factory.go:
```diff
@@ -0,0 +1,120 @@
+
+// extract the port number from string in "address:port" format. If the
+// port number cannot be extracted returns an error.
+func extractPortFromEndpoint(endpoint string) (int, error) {
```
We should probably make this a utility func available for all components. No need to change anything in this PR, maybe just a TODO note for future improvement.

codesmith14 replied:
added a TODO

tigrannajaryan commented on receiver/sapmreceiver/trace_receiver.go:
```diff
@@ -0,0 +1,159 @@
+				td.SourceFormat = "sapm"
+
+				// pass the trace data to the next consumer
+				sr.nextConsumer.ConsumeTraceData(rCtx, td)
```
I think `ConsumeTraceData` can return an error, it needs to be handled here otherwise we may have silent drops.

codesmith14 replied:
@tigrannajaryan Is returning on error appropriate for converting the batches and consuming their trace data?  Or should I attempt to process the rest of the batches before returning?

tigrannajaryan replied:
We have 2 options:
1. Return ErrBadRequest error to the caller. This will ensure caller does not re-send this anymore.
2. Return "Success" but increment local "bad input data" counter and process the rest of batches which are good.

(1) is definitely more appropriate if ProtoBatchToOCProto returns an error because it typically means something is wrong with span data.
(2) _may_ be more appropriate for `ConsumeTraceData` errors, but I could argue either way.

To keep it simple I would handle errors from both functions the same way and would just return ErrBadRequest (but make sure we actually return this exact error so that SAPM responds with 400 Bad Data and we don't receive this data again).

tigrannajaryan commented on receiver/sapmreceiver/trace_receiver.go:
```diff
@@ -0,0 +1,159 @@
+	return func(rw http.ResponseWriter, req *http.Request) {
+
+		// trace this request
+		_, span := trace.StartSpan(sr.defaultCtx, traceSource)
```
@owais @pjanotti Do we have own trace collection enabled in the Collector? Do these spans go anywhere or it is a noop currently?

tigrannajaryan commented on receiver/sapmreceiver/trace_receiver.go:
```diff
@@ -0,0 +1,159 @@
+		// set up the listener
+		ln, err = net.Listen("tcp", sr.config.Endpoint)
+		if err != nil {
+			err = fmt.Errorf("failed to bind to address %s: %v", sr.config.Endpoint, err)
```
Shouldn't we `return` here? Is there a point in continuing and trying to setup the http.Server?

codesmith14 replied:
added a return

tigrannajaryan commented:
Nit: capitalize subject line of commit message.
(At Omnition we were typically following this advice: https://chris.beams.io/posts/git-commit/)

codesmith14 commented:
After discussing with @owais, I'm going to remove the handler from sapm-proto.  I'll leave the request parsing function, but the handler itself will be moved into the otel receiver directly.

codesmith14 commented:
@tigrannajaryan @owais I think I've addressed all comments on this pr.  I'll update the sapm-proto dependency with a tagged version once this pr is approved: https://github.com/signalfx/sapm-proto/pull/7

pjanotti reviewed (commented):
Looks good in general. I miss somewhere a description for SAPM (Splunk APM?), best place will be adding a doc.go for the package and describe it there.

codesmith14 commented on receiver/sapmreceiver/trace_receiver.go:
```diff
@@ -0,0 +1,239 @@
+
+	// pass the SAPM and error to the receiver function
+	// process the sapm
+	if sapm == nil {
```
I guess it can't, I'll remove the nil check

codesmith14 commented on receiver/sapmreceiver/factory.go:
```diff
@@ -0,0 +1,117 @@
+	// assert config is SAPM config
+	rCfg := cfg.(*Config)
+
+	port, err := extractPortFromEndpoint(rCfg.Endpoint)
```
I copied this check from the jaeger receiver.  I think it's primary purpose is to prevent a 0 valued port from being passed to the listener.  This wouldn't raise an error it would just assign a random port to the listener.

pjanotti replied:
hum, perhaps we need a better and general way to handle that. Anyway, it is a separated issue.

codesmith14 commented on receiver/sapmreceiver/trace_receiver.go:
```diff
@@ -0,0 +1,239 @@
+		return
+	}
+
+	// gzip the response
```
I think the thinking is that in the future this might return something.  I'll go ahead and make a default response in the receiver's constructor and send that for now.

owais replied:
I think we should do it if acceptEndocingHeader is present. At least it documents what the behavior should be. As an optimization, until we actually need a real response, we can have the client never ask for gzipped response. Either that or the encoding code below can just check if the response to be sent is empty and skip encoding.

codesmith14 replied:
Ok cool, then I'm just going to have a default response in []byte.  And we'll still pass that through the gzip code here.  If we ever decide to put stuff in the response we can revise this function to actually build the response byte array before going through this gzip path.

codesmith14 commented on receiver/sapmreceiver/trace_receiver.go:
```diff
@@ -0,0 +1,239 @@
+		ln, err = net.Listen("tcp", sr.config.Endpoint)
+		if err != nil {
+			err = fmt.Errorf("failed to bind to address %s: %v", sr.config.Endpoint, err)
+			return
```
The naked return is for the function passed into `sr.startOnce.Do()`, the error that is being set in the line above is declared outside of this inner function.  So ultimately we will return the error and not nil.

pjanotti replied:
ops, missed that

tigrannajaryan reviewed (approved):
LGTM on my part, please address remaining comments from other reviewers.

codesmith14 commented on receiver/sapmreceiver/trace_receiver_test.go:
```diff
@@ -0,0 +1,285 @@
+			args: args{
+				// 1. Create the SAPM receiver aka "server"
+				config: &Config{
+					ReceiverSettings: configmodels.ReceiverSettings{Endpoint: ":7276"},
```
I can't use that because that's in a package internal to `optentelemetry-collector` and isn't in `opentelemetry-collector-contrib`.  But since this is the default endpoint, i'll use the constant in this package.

pjanotti replied:
I will take care of exposing that soon (no need to hold this PR for that).

codesmith14 commented on receiver/sapmreceiver/trace_receiver.go:
```diff
@@ -0,0 +1,239 @@
+	// handle the request payload
+	err := sr.handleRequest(ctx, req)
+	if err != nil {
+		// TODO account for this error (throttled logging or metrics)
```
Will do.  @tigrannajaryan is there some kind of convention in otel for handling this?  I don't think zap has a rate limited logger (at least i couldn't find one in their docs).  I hesitate to put a plain log statement there because log files could blow up from a bad actor.

codesmith14 replied:
hrm I think i found something in zap. They refer to this as "sampling".  I'll have to read their doc further to understand how to use that feature.  I've never used zap before.

tigrannajaryan replied:
I think it is best to record a metric instead of logging. I believe that is how we expose most of bad things happening in Collector that can happen because of input data.

codesmith14 commented:
I'll continue looking into the logging stuff separately and open another pr to address it.  @tigrannajaryan @pjanotti @owais I think this pr is ready to merge pending your final approval.

pjanotti reviewed (approved):
Thanks @codesmith14 LGTM
